openverse_catalog/dags/maintenance/fix_phylopic_foreign_identifier.py

```diff
@@ -0,0 +1,116 @@
+"""
+One-time run DAG to fix the foreign identifier for PhyloPic images.
+
+In order to prevent broken links, we need to update the foreign identifier from using
+the image URL to using the foreign image UUID.
+"""
+import logging
+import re
+from textwrap import dedent
+
+from airflow.decorators import dag
+from airflow.models.abstractoperator import AbstractOperator
+from airflow.operators.python import PythonOperator
+from psycopg2.errors import UniqueViolation
+
+from common.constants import DAG_DEFAULT_ARGS, POSTGRES_CONN_ID, XCOM_PULL_TEMPLATE
+from common.slack import send_message
+from common.sql import PostgresHook
+
+
+logger = logging.getLogger(__name__)
+
+DAG_ID = "update_phylopic_foreign_identifier"
+
+
+def update_foreign_identifiers(task: AbstractOperator) -> dict[str, int]:
+    pg = PostgresHook(
+        postgres_conn_id=POSTGRES_CONN_ID,
+        default_statement_timeout=PostgresHook.get_execution_timeout(task),
+    )
+    phylopic_rows = pg.get_records(
+        "SELECT foreign_identifier, identifier FROM image WHERE provider = 'phylopic';"
+    )
+
+    uuid_pattern = (
+        "[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}"
+    )
+    counter = {
+        "updated": 0,
+        "skipped": 0,
+        "failed_none": 0,
+        "failed_duplicate": 0,
+        "failed_other_reason": 0,
+    }
+
+    for foreign_identifier, identifier in phylopic_rows:
+        if len(foreign_identifier) == 36:
+            # The foreign identifier is (most likely) already the UUID
+            counter["skipped"] += 1
+            continue
+
+        uuid = re.search(uuid_pattern, foreign_identifier).group()
+        if uuid is None:
+            counter["failed_none"] += 1
+            continue
+
+        # Update the foreign identifier with the UUID
+        update_query = dedent(
+            f"""
+            UPDATE image SET foreign_identifier = '{uuid}', thumbnail = NULL
+            WHERE identifier = '{identifier}'
+            """
+        )
+
+        try:
+            pg.run(update_query)
+            counter["updated"] += 1
+        except UniqueViolation:
+            logger.warning(
+                f"Duplicate foreign identifier {uuid} for identifier {identifier}"
+            )
+            counter["failed_duplicate"] += 1
+        except Exception as e:
+            counter["failed_other_reason"] += 1
+            logger.error(
+                f"Failed to update foreign identifier for identifier {identifier}"
+                f" with error: {e}"
+            )
+
+    return counter
+
+
+def final_report(counter: dict[str, int]) -> None:
+    message = f"{DAG_ID} DAG run completed. Update statistics:\n{counter}."
```

Comment:
Since this will be sent to Slack, it might be better to either explicitly print the statistics with keys & values _or_ to wrap `counter` in backticks:

```suggestion
    counter_message = "\n".join([f"{key.split('_', ' ').title()}: {value}" for key, value in counter.items()])
    message = f"{DAG_ID} DAG run completed. Update statistics:\n{counter_message}."
```

Reply:
I tried that but even though the data type says it's a `dict`, for some reason at that point `counter` is a `str` 🤷🏻‍♀️ I'll better remove that type hint to avoid confusion.

Reply:
Oh, it's probably because we're missing this arg:
https://github.com/WordPress/openverse-catalog/blob/3f2ab0967991a45bb82038e05024d7b82fe06ae2/openverse_catalog/dags/database/terminate_long_queries_workflow.py#L75